File: markdown/self-drive-car/src/car.ts
import { drawRect } from "./drawUtils";
import { Rectangle, CarFrameMargin, CarHeight, CarWidth, CarPos, FrameWidth, FrameHeight, CanvasWidth, CanvasHeight, UseBrain, StartX, StartY, RoadMap, EmptyRoad, IRoad, RadarLineLength, objectToArray, ICarState } from "./gameUtils";
import car1 from './assets/car1.png';
import { Controls } from "./controls";
import { ILine, IPosition, IRect, getRectangleWidthHeight, getTwoPointsDistance, rotateRectangle, updateCoordinates } from "./math";
import { Radar } from "./radar";
import { Brain } from "./brain";


export class Car {
    carImage: HTMLImageElement;
    pos: IPosition = { x: 0, y: 0 };
    frame: Rectangle;

    controls: Controls;
    speed = 0;
    acceleration = 0.3;
    maxSpeed = 4;
    friction = 0.02;
    angle = 270;
    damaged = false;
    moveDistance = 0;

    x = 0;
    y = 0;

    radar = new Radar();
    brain: Brain;
    prevState = {
        x: 0,
        y: 0,
        angle: 0,
        moveDistance: 0,
    };

    constructor() {
        this.carImage = new Image();
        this.carImage.src = car1;
        this.frame = new Rectangle(
            {
                x: this.pos.x + CarFrameMargin,
                y: this.pos.y + CarFrameMargin
            },
            CarWidth - CarFrameMargin * 2,
            CarHeight - CarFrameMargin * 2);
        this.controls = new Controls();
        this.brain = new Brain(objectToArray(this.getStateObj()));
    }

    render(ctx: CanvasRenderingContext2D) {
        const x = this.pos.x;
        const y = this.pos.y;
        //ctx.globalCompositeOperation = 'destination-atop';
        const angleInRadians = (this.angle - 270) * (Math.PI / 180);
        // 將原點移至圖像的中心
        ctx.translate(x + CarWidth / 2, y + CarHeight / 2);
        ctx.rotate(angleInRadians);
        ctx.drawImage(this.carImage, -CarWidth / 2, -CarHeight / 2);
        ctx.setTransform(1, 0, 0, 1, 0, 0);

        //console.log(`car ${this.x} ${this.y}`);
        const radar = this.radar;
        radar.carXY = { x: this.x, y: this.y };
        radar.carAngle = this.angle;
        radar.pos = {
            x: this.pos.x + CarWidth / 2,
            y: this.pos.y + CarHeight / 2,
        }
        radar.render(ctx);

        // 印出 sensor 數字
        //const sensors = [];
        //for(let radarLine of radar.radarLines){
        //    sensors.push(radarLine.distance);
        //}
        //console.log(`sensors ${sensors.join(', ')}`)
    }

    getBoundLines(): ILine[] {
        const { leftTop, rightTop, rightBottom, leftBottom } = this.getBound();
        return [
            { start: leftTop, end: rightTop },
            { start: rightTop, end: rightBottom },
            { start: rightBottom, end: leftBottom },
            { start: leftBottom, end: leftTop },
        ];
    }

    getBound(): IRect {
        let x1 = this.x - FrameWidth / 2;
        let y1 = this.y - FrameHeight / 2;
        let x2 = x1 + FrameWidth;
        let y2 = y1 + FrameHeight;
        const [leftTop, rightTop, rightBottom, leftBottom] = rotateRectangle({ x: x1, y: y1 }, { x: x2, y: y2 }, this.angle);
        return { leftTop, rightTop, rightBottom, leftBottom, }
    }

    getFrame(): IRect {
        let x1 = CanvasWidth / 2 - FrameWidth / 2;
        let y1 = CanvasHeight / 2 - FrameHeight / 2;
        let x2 = x1 + FrameWidth;
        let y2 = y1 + FrameHeight;
        const [leftTop, rightTop, rightBottom, leftBottom] = rotateRectangle(
            { x: x1, y: y1 },
            { x: x2, y: y2 },
            this.angle);
        return {
            leftTop,
            rightTop,
            rightBottom,
            leftBottom
        };
    }

    getDrawBound(): IRect {
        let x1 = CanvasWidth / 2 - FrameWidth / 2;
        let y1 = CanvasHeight / 2 - FrameHeight / 2;
        let x2 = x1 + FrameWidth;
        let y2 = y1 + FrameHeight;
        const [leftTop, rightTop, rightBottom, leftBottom] = rotateRectangle({ x: x1, y: y1 }, { x: x2, y: y2 }, this.angle);
        return { leftTop, rightTop, rightBottom, leftBottom, }
    }

    drawFrame(ctx: CanvasRenderingContext2D) {
        drawRect(ctx, this.getFrame(), { lineWidth: 5, strokeSyle: "yellow" });
        //const carBound = this.getBound();
        // const { leftTop: p0, rightTop: p1, rightBottom: p2, leftBottom: p3 } = this.getDrawBound();
        // drawText(ctx, p0, `${posInfo(carBound.leftTop)}`)
        // drawText(ctx, p1, `${posInfo(carBound.rightTop)}`)
        // drawText(ctx, p2, `${posInfo(carBound.leftBottom)}`)
        // drawText(ctx, p3, `${posInfo(carBound.rightBottom)}`)
    }

    async move(ctx: CanvasRenderingContext2D, roadMap: RoadMap) {
        const pos0 = { x: this.x, y: this.y };
        this.prevState.x = this.x;
        this.prevState.y = this.y;
        this.prevState.angle = this.angle;
        this.prevState.moveDistance = this.moveDistance;

        const brain = this.brain;
        if (UseBrain) {
            const self = this;
            const action = await brain.control(self.getState.bind(self));
            this.controls.forward = false;
            this.controls.reverse = false;
            this.controls.left = false;
            this.controls.right = false;
            switch (action) {
                case 0:
                    this.controls.forward = true;
                    break;
                case 1:
                    this.controls.left = true;
                    break;
                case 2:
                    this.controls.right = true;
                    break;
                case 3:
                    this.controls.reverse = true;
                    break;
            }
            //console.log('brain', action);
        }

        if (this.controls.forward) {
            this.speed += this.acceleration;
        }

        if (this.controls.reverse) {
            this.speed -= this.acceleration;
        }

        if (this.speed > this.maxSpeed) {
            this.speed = this.maxSpeed;
        }

        if (this.speed < -this.maxSpeed / 2) {
            this.speed = -this.maxSpeed / 2;
        }

        //自動減速效果
        if (this.speed > 0) {
            this.speed -= this.friction;
        }
        if (this.speed < 0) {
            this.speed += this.friction;
        }
        if (Math.abs(this.speed) < this.friction) {
            this.speed = 0;
        }

        if (this.controls.left) {
            this.angle -= 1;
        }

        if (this.controls.right) {
            this.angle += 1;
        }

        const pos1 = updateCoordinates(pos0, this.angle, this.speed);
        this.x = pos1.x;
        this.y = pos1.y;
        if( this.speed > 0){
            this.moveDistance += this.speed;
        }

        await this.collideAsync(ctx, roadMap);
        this.renderRadarLines(ctx, roadMap);
        return [pos0, pos1];
    }

    renderRadarLines(ctx: CanvasRenderingContext2D, roadMap: RoadMap) {
        const radars = this.radar;
        for (let [index, radarLine] of radars.getBoundLines().entries()) {
            const radar = radars.radarLines[index];
            const [road, collideRadarPoints] = roadMap.collide(ctx, [radarLine]);
            if (collideRadarPoints.length != 0) {
                radar.renderDamaged(ctx, collideRadarPoints[0]);
            } else {
                radar.distance = RadarLineLength;
            }
        }
    }

    getStateObj(): ICarState {
        const radarSense = this.radar.radarLines.map(x => x.distance);
        return {
            x: this.x,
            y: this.y,
            angle: this.angle,
            speed: this.speed,
            damaged: this.damaged ? 1 : 0,
            moveDistance: this.moveDistance,
            radarSense
        }
    }

    getState(): number[] {
        const state = this.getStateObj();
        return objectToArray(state).values;
    }

    async collideAsync(ctx: CanvasRenderingContext2D, roadMap: RoadMap): Promise<[IRoad, IPosition[]]> {
        const boundLines = this.getBoundLines();
        const roads = roadMap.roads;
        for (let ix = 0; ix < roads.length; ix++) {
            for (let iy = 0; iy < roads[ix].length; iy++) {
                const road = roads[ix][iy];
                const collidePoints = road.collide(ctx, boundLines);
                if (collidePoints.length > 0) {
                    this.x = this.prevState.x;
                    this.y = this.prevState.y;
                    this.angle = this.prevState.angle;
                    this.moveDistance = this.prevState.moveDistance;
                    this.speed = 0;
                    this.damaged = true;
                    road.renderDamaged(ctx);

                    if (UseBrain) {
                        await this.brain.control(this.getState.bind(this));
                        this.x = StartX;
                        this.y = StartY;
                        this.angle = 270;
                        this.speed = 0;
                        this.damaged = false;
                        this.moveDistance = 0;
                    }

                    return [road, collidePoints];
                }
            }
        }
        this.damaged = false;
        return [EmptyRoad.Default, []];
    }
}

File: torch-self-drive-car/car.py
import math
from enum import Enum
import copy

from constants import FrameWidth, FrameHeight, CenterX, CenterY, StartX, StartY
from radar import Radar
from roads import EmptyRoad, RoadMap
from math_utils import Position, update_coordinates, Line, rotate_rectangle
from pygameGraphic import IGraphic, PygameController


class Action(Enum):
    NONE = 0
    UP = 1
    LEFT = 2
    RIGHT = 3
    DOWN = 4


class CarState:
    def __init__(self):
        self.speed = 0
        self.angle = 0
        self.radar_lines = []

    def to_dict(self):
        return self.__dict__

    def clone(self):
        return copy.copy(self)


class Car:
    pos = Position(0, 0)
    controller = PygameController()
    x = 0
    y = 0
    prev_x = 0
    prev_y = 0
    state = CarState()
    prev_state = CarState()

    def __init__(self):
        self.radar = Radar()
        self.controller.create()
        self.acceleration = 0.3
        self.max_speed = 4
        self.friction = 0.02
        self.damaged = False

    def reset(self):
        self.x = StartX
        self.y = StartY
        self.state.speed = 0
        self.state.angle = 0
        self.damaged = False

    def render(self, ctx: IGraphic):
        state = self.state
        radar = self.radar
        ctx.draw_image("./assets/car1.png", self.pos, state.angle)
        radar.pos = self.pos
        radar.car_xy = Position(self.x, self.y)
        radar.car_angle = state.angle
        radar.render(ctx)
        bound_line = self.get_frame_lines()
        for line in bound_line:
            start = Position(line.start.x, line.start.y)
            end = Position(line.end.x, line.end.y)
            ctx.draw_line(Line(start, end), color="yellow", thickness=5)
        self.controller.render()

    def get_observation_info(self) -> CarState:
        info = self.state.clone()
        info.radar_lines = self.radar.get_observation_info()
        return info

    def control(self, action: Action):
        if action == Action.UP:
            self.controller.forward = True
        if action == Action.DOWN:
            self.controller.reverse = True
        if action == Action.LEFT:
            self.controller.left = True
        if action == Action.RIGHT:
            self.controller.right = True
        if action == Action.NONE:
            self.controller.forward = False
            self.controller.reverse = False
            self.controller.left = False
            self.controller.right = False

    def rollback_state(self):
        self.state = self.prev_state

    def move(self, ctx: IGraphic, road_map: RoadMap):
        state = self.state
        if self.controller.forward:
            state.speed += self.acceleration
        elif self.controller.reverse:
            state.speed -= self.acceleration

        if state.speed > self.max_speed:
            state.speed = self.max_speed
        elif state.speed < -self.max_speed / 2:
            state.speed = -self.max_speed / 2

        if state.speed > 0:
            state.speed -= self.friction
        elif state.speed < 0:
            state.speed += self.friction
        if abs(state.speed) < self.friction:
            state.speed = 0

        if self.controller.left:
            state.angle += 1
        elif self.controller.right:
            state.angle -= 1

        new_pos = update_coordinates(Position(self.x, self.y), state.angle, state.speed)
        self.x = new_pos.x
        self.y = new_pos.y

        self.collide(ctx, road_map)

    def collide(self, ctx: IGraphic, road_map: RoadMap):
        self.radar.collide(ctx, road_map)
        bound_lines = self.get_bound_lines()
        roads = road_map.roads
        for ix in range(len(roads)):
            for iy in range(len(roads[ix])):
                road = roads[ix][iy]
                collide_points = road.collide(ctx, bound_lines)
                if len(collide_points) > 0:
                    # print(f"BOOM")
                    self.x = self.prev_x
                    self.y = self.prev_y
                    self.state = self.prev_state.clone()
                    self.state.speed = 0
                    self.damaged = True
                    road.render_damaged(ctx)
                    return [road, collide_points]
        self.prev_x = self.x
        self.prev_y = self.y
        self.prev_state = self.state.clone()
        self.prev_state.radar_lines = self.radar.get_observation_info()
        return [EmptyRoad(), []]

    def get_bound_lines(self) -> list[Line]:
        [left_top, right_top, right_bottom, left_bottom] = self.get_bound_points()
        return [
            Line(left_top, right_top),
            Line(right_top, right_bottom),
            Line(right_bottom, left_bottom),
            Line(left_bottom, left_top)
        ]

    def get_bound_points(self):
        state = self.state
        left_top = Position(self.x - FrameWidth / 2, self.y - FrameHeight / 2)
        right_bottom = Position(left_top.x + FrameWidth, left_top.y + FrameHeight)
        return rotate_rectangle(left_top, right_bottom, state.angle)

    def get_frame_lines(self):
        left_top = Position(self.pos.x - FrameWidth / 2, self.pos.y - FrameHeight / 2)
        right_bottom = Position(left_top.x + FrameWidth, left_top.y + FrameHeight)
        [left_top, right_top, right_bottom, left_bottom] = rotate_rectangle(left_top, right_bottom, self.state.angle)
        return [
            Line(left_top, right_top),
            Line(right_top, right_bottom),
            Line(right_bottom, left_bottom),
            Line(left_bottom, left_top)
        ]
